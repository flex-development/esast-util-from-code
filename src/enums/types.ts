/**
 * @file Enums - types
 * @module esast-util-from-code/enums/types
 */

import type { EsastNode } from '@flex-development/esast'
import type { Type } from '@flex-development/unist-util-types'

/**
 * esast node types.
 *
 * @enum {Type<EsastNode>}
 */
enum types {
  ambientDeclaration = 'ambientDeclaration',
  argumentList = 'argumentList',
  arithmeticExpression = 'arithmeticExpression',
  arrayExpression = 'arrayExpression',
  arrayPattern = 'arrayPattern',
  arrayType = 'arrayType',
  arrowFunctionExpression = 'arrowFunctionExpression',
  asClause = 'asClause',
  asExpression = 'asExpression',
  assertionPredicate = 'assertionPredicate',
  assignmentExpression = 'assignmentExpression',
  assignmentPattern = 'assignmentPattern',
  awaitExpression = 'awaitExpression',
  bigint = 'bigint',
  bitwiseExpression = 'bitwiseExpression',
  blockStatement = 'blockStatement',
  boolean = 'boolean',
  breakStatement = 'breakStatement',
  callExpression = 'callExpression',
  callSignature = 'callSignature',
  catchClause = 'catchClause',
  classBody = 'classBody',
  classDeclaration = 'classDeclaration',
  classExpression = 'classExpression',
  classHeritage = 'classHeritage',
  comment = 'comment',
  computedExpression = 'computedExpression',
  conditionalExpression = 'conditionalExpression',
  conditionalType = 'conditionalType',
  constructSignature = 'constructSignature',
  constructorType = 'constructorType',
  continueStatement = 'continueStatement',
  debuggerStatement = 'debuggerStatement',
  decorator = 'decorator',
  decoratorList = 'decoratorList',
  directive = 'directive',
  doWhileStatement = 'doWhileStatement',
  elseStatement = 'elseStatement',
  emptyStatement = 'emptyStatement',
  enumBody = 'enumBody',
  enumDeclaration = 'enumDeclaration',
  enumMember = 'enumMember',
  equalityExpression = 'equalityExpression',
  exportDeclaration = 'exportDeclaration',
  exportSpecifier = 'exportSpecifier',
  exportSpecifiers = 'exportSpecifiers',
  expressionStatement = 'expressionStatement',
  extendsClause = 'extendsClause',
  fieldDefinition = 'fieldDefinition',
  finallyBlock = 'finallyBlock',
  forInStatement = 'forInStatement',
  forOfStatement = 'forOfStatement',
  forStatement = 'forStatement',
  fromClause = 'fromClause',
  functionDeclaration = 'functionDeclaration',
  functionExpression = 'functionExpression',
  functionSignature = 'functionSignature',
  functionType = 'functionType',
  genericType = 'genericType',
  identifier = 'identifier',
  ifStatement = 'ifStatement',
  implementsClause = 'implementsClause',
  importAssertion = 'importAssertion',
  importAttribute = 'importAttribute',
  importDeclaration = 'importDeclaration',
  importExpression = 'importExpression',
  importMeta = 'importMeta',
  importSpecifier = 'importSpecifier',
  importSpecifiers = 'importSpecifiers',
  indexSignature = 'indexSignature',
  indexedAccessType = 'indexedAccessType',
  inferType = 'inferType',
  interfaceBody = 'interfaceBody',
  interfaceDeclaration = 'interfaceDeclaration',
  intersectionType = 'intersectionType',
  labeledStatement = 'labeledStatement',
  logicalExpression = 'logicalExpression',
  mappedSignature = 'mappedSignature',
  mappedType = 'mappedType',
  memberExpression = 'memberExpression',
  methodDefinition = 'methodDefinition',
  methodSignature = 'methodSignature',
  modifier = 'modifier',
  modifierList = 'modifierList',
  moduleBody = 'moduleBody',
  moduleDeclaration = 'moduleDeclaration',
  moduleExpression = 'moduleExpression',
  namespaceDeclaration = 'namespaceDeclaration',
  namespaceExport = 'namespaceExport',
  namespaceImport = 'namespaceImport',
  newExpression = 'newExpression',
  nonNullExpression = 'nonNullExpression',
  nothing = 'nothing',
  null = 'null',
  number = 'number',
  objectExpression = 'objectExpression',
  objectLiteralType = 'objectLiteralType',
  objectPattern = 'objectPattern',
  parameter = 'parameter',
  parameterList = 'parameterList',
  parenthesizedExpression = 'parenthesizedExpression',
  parenthesizedType = 'parenthesizedType',
  property = 'property',
  propertySignature = 'propertySignature',
  regexp = 'regexp',
  relationalExpression = 'relationalExpression',
  restElement = 'restElement',
  restType = 'restType',
  returnStatement = 'returnStatement',
  root = 'root',
  satisfiesClause = 'satisfiesClause',
  satisfiesExpression = 'satisfiesExpression',
  sequenceExpression = 'sequenceExpression',
  spreadElement = 'spreadElement',
  staticBlock = 'staticBlock',
  staticBlockBody = 'staticBlockBody',
  string = 'string',
  super = 'super',
  switchBody = 'switchBody',
  switchCase = 'switchCase',
  switchStatement = 'switchStatement',
  taggedTemplateExpression = 'taggedTemplateExpression',
  templateElement = 'templateElement',
  templateLiteral = 'templateLiteral',
  templatePlaceholder = 'templatePlaceholder',
  this = 'this',
  throwStatement = 'throwStatement',
  tryStatement = 'tryStatement',
  tupleElement = 'tupleElement',
  tupleType = 'tupleType',
  typeAliasDeclaration = 'typeAliasDeclaration',
  typeAnnotation = 'typeAnnotation',
  typeArgumentList = 'typeArgumentList',
  typeAssertion = 'typeAssertion',
  typeAssertionExpression = 'typeAssertionExpression',
  typeKeyword = 'typeKeyword',
  typeParameter = 'typeParameter',
  typeParameterList = 'typeParameterList',
  typePredicate = 'typePredicate',
  unaryExpression = 'unaryExpression',
  unaryTypeExpression = 'unaryTypeExpression',
  undefined = 'undefined',
  unionType = 'unionType',
  updateExpression = 'updateExpression',
  variableDeclaration = 'variableDeclaration',
  variableDeclarator = 'variableDeclarator',
  whileStatement = 'whileStatement',
  withStatement = 'withStatement',
  yieldExpression = 'yieldExpression'
}

export default types
