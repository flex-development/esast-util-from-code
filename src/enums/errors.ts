/**
 * @file Enums - errors
 * @module esast-util-from-code/enums/errors
 */

/**
 * Error messages.
 *
 * @enum {string}
 */
enum errors {
  comment = 'Unterminated comment',
  string = 'Unterminated string constant'
}

export default errors
