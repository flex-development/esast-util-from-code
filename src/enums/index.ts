/**
 * @file Entry Point - Enums
 * @module esast-util-from-code/enums
 */

export { default as errors } from './errors'
export { default as tt } from './tt'
