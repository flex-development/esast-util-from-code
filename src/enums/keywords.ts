/**
 * @file Enums - keywords
 * @module esast-util-from-code/enums/keywords
 */

/**
 * Possible keywords.
 *
 * @enum {Lowercase<string>}
 */
enum keywords {
  abstract = 'abstract',
  accessor = 'accessor',
  any = 'any',
  as = 'as',
  assert = 'assert',
  asserts = 'asserts',
  async = 'async',
  await = 'await',
  bigint = 'bigint',
  boolean = 'boolean',
  break = 'break',
  case = 'case',
  catch = 'catch',
  class = 'class',
  const = 'const',
  constructor = 'constructor',
  continue = 'continue',
  debugger = 'debugger',
  declare = 'declare',
  default = 'default',
  delete = 'delete',
  do = 'do',
  else = 'else',
  enum = 'enum',
  export = 'export',
  extends = 'extends',
  false = 'false',
  finally = 'finally',
  for = 'for',
  from = 'from',
  function = 'function',
  get = 'get',
  if = 'if',
  implements = 'implements',
  import = 'import',
  in = 'in',
  infer = 'infer',
  instanceof = 'instanceof',
  interface = 'interface',
  keyof = 'keyof',
  let = 'let',
  meta = 'meta',
  module = 'module',
  namespace = 'namespace',
  never = 'never',
  new = 'new',
  null = 'null',
  number = 'number',
  object = 'object',
  of = 'of',
  out = 'out',
  override = 'override',
  private = 'private',
  protected = 'protected',
  public = 'public',
  readonly = 'readonly',
  return = 'return',
  satisfies = 'satisfies',
  set = 'set',
  static = 'static',
  string = 'string',
  super = 'super',
  switch = 'switch',
  symbol = 'symbol',
  this = 'this',
  throw = 'throw',
  true = 'true',
  try = 'try',
  type = 'type',
  typeof = 'typeof',
  undefined = 'undefined',
  unique = 'unique',
  unknown = 'unknown',
  using = 'using',
  var = 'var',
  void = 'void',
  while = 'while',
  with = 'with',
  yield = 'yield'
}

export default keywords
