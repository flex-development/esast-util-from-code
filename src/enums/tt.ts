/**
 * @file Enums - tt
 * @module esast-util-from-code/enums/tt
 */

import type { TokenType } from '@flex-development/vfile-lexer'

/**
 * Token kinds.
 *
 * @see {@linkcode TokenType}
 *
 * @enum {TokenType}
 */
enum tt {
  bigint = 'bigint',
  comment = 'comment',
  eof = 'eof',
  keyid = 'keyid',
  number = 'number',
  punctuator = 'punctuator',
  sof = 'sof',
  string = 'string',
  whitespace = 'whitespace'
}

export default tt
