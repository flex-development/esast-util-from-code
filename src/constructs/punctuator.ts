/**
 * @file Constructs - punctuator
 * @module esast-util-from-code/constructs/punctuator
 */

import { tt } from '#src/enums'
import {
  resolveSlice,
  type Code,
  type Construct,
  type Effects,
  type Guard,
  type Resolver,
  type State,
  type TokenizeContext
} from '@flex-development/vfile-lexer'
import { ok as assert } from 'devlop'
import { unicodePunctuation as test } from 'micromark-util-character'

/**
 * Punctuator construct.
 *
 * @const {Construct} punctuator
 */
const punctuator: Construct = {
  /**
   * Construct name.
   */
  name: tt.punctuator,

  /**
   * Resolve the events parsed by `tokenize`.
   *
   * @see {@linkcode Resolver}
   */
  resolve: resolveSlice,

  /**
   * Check if the current character code can start this construct.
   *
   * @see {@linkcode Guard}
   */
  test,

  /**
   * Set up a state machine to handle character codes streaming in.
   *
   * @see {@linkcode Effects}
   * @see {@linkcode State}
   * @see {@linkcode TokenizeContext}
   *
   * @this {TokenizeContext}
   *
   * @param {Effects} effects - Context object to transition state machine
   * @param {State} ok - Successful tokenization state
   * @return {State} Initial state
   */
  tokenize(this: TokenizeContext, effects: Effects, ok: State): State {
    /**
     * Tokenize context.
     *
     * @const {TokenizeContext} self
     */
    const self: TokenizeContext = this

    return punctuator

    /**
     * Tokenize a punctuator.
     *
     * @param {Code} code - Current character code
     * @return {State} Next state
     */
    function punctuator(code: Code): State {
      assert(test.call(self, code), 'expected code to pass test')
      effects.enter(tt.punctuator)
      effects.consume(code)
      effects.exit(tt.punctuator)
      return ok
    }
  }
}

export default punctuator
