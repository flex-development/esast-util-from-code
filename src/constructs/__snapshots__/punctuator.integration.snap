// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:constructs/punctuator > should tokenize chars.ampersand 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "&" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.asterisk 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "*" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.at 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "@" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.backslash 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "\\\\" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.backtick 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "\`" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.bar 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "|" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.caret 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "^" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.colon 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator ":" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.comma 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "," (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.dot 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "." (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.equal 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "=" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.exclamation 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "!" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.gt 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator ">" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.hash 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "#" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.leftBrace 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "{" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.leftBracket 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "[" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.leftParen 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "(" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.lt 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "<" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.minus 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "-" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.percent 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "%" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.plus 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "+" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.question 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "?" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.rightBrace 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "}" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.rightBracket 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "]" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.rightParen 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator ")" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.semicolon 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator ";" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.slash 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "/" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/punctuator > should tokenize chars.tilde 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 punctuator "~" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;
